src/StateMaps/PDEConstrainedFunctionals.jl

```diff
@@ -428,3 +428,127 @@ function Base.show(io::IO,::MIME"text/plain",f::AbstractPDEConstrainedFunctional
   print(io,"$(nameof(typeof(f))):
     num_constraints: $N")
 end
+
+struct CustomPDEConstrainedFunctionals{N,A} <:  AbstractPDEConstrainedFunctionals{N}
+  φ_to_jc :: Function
+  dJ :: Vector{Float64}
+  dC :: Vector{Vector{Float64}}
+  analytic_dJ
+  analytic_dC
+  state_map :: A
+
+    function CustomPDEConstrainedFunctionals(
+      φ_to_jc :: Function,
+      state_map :: AbstractFEStateMap,
+      φh;
+    )
+
+    # Pre-allocaitng
+    grad = Zygote.jacobian(φ_to_jc, φh.free_values)
+    dJ = grad[1][1,:]
+    dC = [collect(row) for row in eachrow(grad[1][2:end,:])]    
+
+    N = length(dC)
+    A = typeof(state_map)
+    analytic_dJ = nothing
+    analytic_dC = fill(nothing,N)
+
+    return new{N,A}(φ_to_jc,dJ,dC,analytic_dJ,analytic_dC,state_map)
+  end
+end
+
+function Fields.evaluate!(pcf::CustomPDEConstrainedFunctionals,φh)
+  φ_to_jc,dJ,dC = pcf.φ_to_jc,pcf.dJ,pcf.dC
+
+  obj,grad = Zygote.withjacobian(φ_to_jc, φh.free_values)
+  j = obj[1]
+  c = obj[2:end]
+  copy!(dJ,grad[1][1,:])
+  copy!(dC,[collect(row) for row in eachrow(grad[1][2:end,:])])
+
+  return j,c,dJ,dC
+end
+
+get_state(m::CustomPDEConstrainedFunctionals) = get_state(m.state_map)
+get_state_map(m::CustomPDEConstrainedFunctionals) = m.state_map
+
+function evaluate_functionals!(pcf::CustomPDEConstrainedFunctionals,φh::FEFunction)
+  φ = φh.free_values
+  return evaluate_functionals!(pcf,φ)
+end
+
+function evaluate_functionals!(pcf::CustomPDEConstrainedFunctionals,φ::AbstractVector)
+  φ_to_jc =  pcf.φ_to_jc
+  obj = φ_to_jc(φ)
+  j = obj[1]
+  c = obj[2:end]
+  return j,c
+end
+
+struct CustomEmbeddedPDEConstrainedFunctionals{N,A} <:  AbstractPDEConstrainedFunctionals{N}

```

Comment:
I think we should try and combine `CustomEmbeddedPDEConstrainedFunctionals` and `CustomPDEConstrainedFunctionals` at some point. Similarly with the `EmbeddedPDEConstrainedFunctionals` and `PDEConstrainedFunctionals`. I think this can come later.

Reply:
totally agree.. even perhaps only have 1 for ALL (no diff btwn custom and not custom since the current CustomEmbeddedPDEConstrainedFunctional works in general) unless there is some benifit in having the specialisation

Reply:
The only issue I see is that we have the state map inside `embedded_collection` but unifying the structs would probably mean we have both as fields. I.e., `state_map::AbstractFEStateMap` and `embedded_collection::Union{EmbeddedCollection,Nothing}`, but the latter could contain a stale version of `state_map`